component/componenttest/lifecycle_verifier.go

```diff
@@ -0,0 +1,79 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package componenttest
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+)
+
+// GetExtensionConfigFn is used customize the configuration passed to the verification.
+// This is used to change ports or provide values required but not provided by the
+// default configuration.
+type GetExtensionConfigFn func() configmodels.Extension
+
+// VerityExtensionLifecycle is used to test if an extension type can handle the typical
+// lifecycle of a component. The getConfigFn parameter only need to be specified if
+// the test can't be done with the default configuration for the component.
+func VerityExtensionLifecycle(t *testing.T, factory component.ExtensionFactory, getConfigFn GetExtensionConfigFn) {
+	ctx := context.Background()
+	host := NewAssertNoError(t)
+	extCreateParams := component.ExtensionCreateParams{
+		Logger:               zap.NewNop(),
+		ApplicationStartInfo: component.DefaultApplicationStartInfo(),
+	}
+
+	var activeExt, builtExt component.Extension
+	defer func() {
+		// If the shutdown happens here there were already some errors on the test.
+		// Ignore errors on this attempt to clean-up.
```

Comment:
Why? Can we just check that there are no errors?

Reply:
If it fails on the require check inside the loop this is a last attempt to clean up resources. Note that in the success case these are never called because both pointers will be nil in the end.

Reply:
Not needed any more with the loop gone https://github.com/open-telemetry/opentelemetry-collector/pull/2679/commits/cface3beec7eedefba00359b7344f1da520ca2e0